File: find_median_data_stream.cpp
#include "leetcode.h"

class MedianFinder {
public:
  void addNum(int num) {
    small.push(num);
    large.push(-small.top());
    small.pop();
    if (small.size() < large.size()) {
      small.push(-large.top());
      large.pop();
    }
  }
  double findMedian() {
    if (small.size() > large.size())
      return small.top();
    else
      return (small.top() - large.top()) / 2.0;
  }

private:
  priority_queue<long> small, large;
};

int main() {
  MedianFinder *obj = new MedianFinder();
  obj->addNum(69);
  double param_2 = obj->findMedian();
}

File: other/find_median_data_stream.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;
#[derive(Default)]
struct MedianFinder {
    small: BinaryHeap<i32>,
    large: BinaryHeap<Reverse<i32>>,
    odd: bool
}

impl MedianFinder {
    fn new() -> Self {
        Default::default()
    }
    fn add_num(&mut self, num: i32) {
        if self.odd {
            self.small.push(num);
            if let Some(max) = self.small.pop() {
                self.large.push(Reverse(max));
            }
        } else {
            self.large.push(Reverse(num));
            if let Some(Reverse(min)) = self.large.pop() {
                self.small.push(min);;
            }
        }
        self.odd = !self.odd;
    }
    fn find_median(&self) -> f64 {
        if self.odd {
            *self.small.peek().unwrap() as f64
        } else {
            (self.small.peek().unwrap() + self.large.peek().unwrap().0) as f64 / 2.0
        }
    }
}

fn main() {
    let obj = MedianFinder::new();
    obj.add_num(num);
    let ret_2: f64 = obj.find_median();
}
